models/stock_move.py

```diff
@@ -37,3 +37,50 @@ def get_info(self):
             res.append(move._prepare_info())
 
         return res
+
+    def _make_mls_comparison_lambda(self, move_line):
+        """ This makes the a lambda for
+            checking the a move_line
+            against move_orign_ids
+        """
+        lot_name = move_line.lot_id.name or move_line.lot_name
+        package = move_line.package_id
+        #lot and package
+        if lot_name and package:
+            return lambda ml: (ml.lot_name == lot_name or \
+                               ml.lot_id.name == lot_name) and \
+                               ml.result_package_id == package
+        # serial
+        elif lot_name:
+            return lambda ml: ml.lot_name == lot_name or \
+                              ml.lot_id.name == lot_name
+
+        # package
+        elif move_line.package_id:
+            return lambda ml: ml.result_package_id == package
+
+        # products
+        else:
+            # This probaly isn't to be trusted
+            return lambda ml: ml.location_dest_id == move_line.location_id and \
+                              ml.product_id == move_line.product_id  # and \
+                              # ml.qty_done <= move_line.ordered_qty
+                              # not sure if the qty comparison makes sense
+
+    def update_orig_ids(self, origin_ids):
+        """ Updates move_orig_ids
+            based on a given set of
+            origin_ids for moves in
+            self.
+        """
+        for move in self:
+            # Retain incomplete moves
+            updated_origin_ids = move.mapped('move_orig_ids').filtered(
+                                            lambda x: x.state not in ('done', 'cancel')
+                                            )
+            for move_line in move.move_line_ids:
+                previous_mls = origin_ids.mapped('move_line_ids').filtered(
```

Comment:
Can we call `origin_ids.mapped('move_line_ids')` once for all instead of `n * m` times? 
I don't think we're changing `origin_ids`, no?